src/Serialization/Fields.jl

```diff
@@ -48,3 +48,132 @@ function load_internal(s::DeserializerState, ::Type{gfp_fmpz_elem}, dict::Dict)
+    K = parent(k)
+    polynomial = parent(defining_polynomial(K))(k)
+    K_dict = save_type_dispatch(s, K)
+
+    return Dict(
+        :parent => K_dict,
+        :polynomial => save_type_dispatch(s, polynomial)
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       ::Type{<: Union{nf_elem, fq_nmod, Hecke.NfRelElem}},
+                       dict::Dict)
+    K = load_type_dispatch(s, dict[:parent], check_namespace=false)
+    polynomial = load_type_dispatch(s, dict[:polynomial], check_namespace=false)
+    return K(polynomial)
+end
+
+################################################################################
+# Non Simple Extension
+function save_internal(s::SerializerState, K::Union{NfAbsNS, NfRelNS})
+    def_pols = defining_polynomials(K)
+    return Dict(
+        :def_pols => save_type_dispatch(s, def_pols),
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       ::Type{<: Union{NfAbsNS, NfRelNS}},
+                       dict::Dict)
+    def_pols = load_type_dispatch(s, dict[:def_pols], check_namespace=false)
+    K, _ = NumberField(def_pols, cached=false)
+    return K
+end
+
+#elements
+function save_internal(s::SerializerState, k::Union{NfAbsNSElem, Hecke.NfRelNSElem})
+    K = parent(k)
+    polynomial = Oscar.Hecke.data(k)
+    polynomial_parent = parent(polynomial)
+    K_dict = save_type_dispatch(s, K)
+    return Dict(
+        :parent_field => K_dict,
+        :polynomial => save_type_dispatch(s, polynomial),
+        :polynomial_parent => save_type_dispatch(s, polynomial_parent)
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       ::Type{<: Union{NfAbsNSElem, Hecke.NfRelNSElem}},
+                       dict::Dict)
+    K = load_type_dispatch(s, dict[:parent_field]; check_namespace=false)
+    polynomial = load_type_dispatch(s, dict[:polynomial]; check_namespace=false)
+    polynomial = evaluate(polynomial, gens(K))
+
+    return K(polynomial)
+end
+
+
+################################################################################
+# FracField
+function save_internal(s::SerializerState, K::FracField)
+    return Dict(
+        :base_ring => save_type_dispatch(s, base_ring(K)),
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       ::Type{<: FracField},
+                       dict::Dict)
+    R, _ = load_type_dispatch(s, dict[:base_ring], check_namespace=false)
+
+    return FractionField(R, cached=false)
+end
+
+# elements
+function save_internal(s::SerializerState, f::FracElem)
+    parent_dict = save_type_dispatch(s, parent(f))
+    return Dict(
+        :parent => parent_dict,
+        :den => save_type_dispatch(s, denominator(f)),
+        :num => save_type_dispatch(s, numerator(f))
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       ::Type{<: FracElem},
+                       dict::Dict)
+    R = load_type_dispatch(s, dict[:parent], check_namespace=false)
```

Comment:
Perhaps could be made explicit via `where T` above?

I am not sure we *want* to move in this direction where we make more things explicit. I see arguments for it (type stability, additional consistency checks, in some cases it may only be possible to proceed with precise types; ...) but also against it (relying on "generic" behavior makes it easier to write serialization methods; may make us more flexible; ...)

So perhaps something to debate, also with others?

Reply:
I think it could be a good idea for now to go the easy route where we just try to serialize as much as we can with the least amount of code. Then when we have uses cases from people trying to de/serialize where the code is taking too long, enforce precise types to get a speed up?

Reply:
The usual gather meeting 1300-1500 on Wednesday?

Reply:
Sure!

Reply:
A few quick comments:

> Here is an example to illustrate why I think this difference matters, and will possibly matter more in the future: currently, `Int(5)` and `fmpz(5)` are serialized in an almost identical fashion, only the `type` string of course is different, `"Int"` (or rather: `"Base.int"`, I think) versus `fmpz`. But from the perspective of a language like Python where there is a single integer type that can store bigints, this distinction makes no sense.

Here we are dealing with a consequence of previous design decisions: it is currently not true (and maybe not intended) that the Oscar types reflect mathematical truths.  To unify ints is generally good idea.  But in view of the current setup it seems better to stick to the Oscar types for now.  Then we can say that all of these represent integers.  Removing the type information now would be a burden, which possibly can be lifted later.

[snip]

> In any case, as I suggested elsewhere, I would at least rename one of the two `load_type_dispatch` methods to a different name to make their difference clearer:

Good idea.

[snip]

> (oohhh, and by the way, right now the first method always returns something of type `T`, _except_ when backreferences are involved, then it could return something of a different type, which sounds like a footgun waiting to be used).

OK.
 
> I suggest we talk about this more on gather. Unfortunately I am busy today and tomorrow, but perhaps Wednesday or Thursday can be arranged?

I could do 14:00 tomorrow (Wed 01 June).

Reply:
I have to apologize but I have to cancel on short notice. An important matter came up I have to deal with immediately.

Sorry!

I'll try to log on gather as soon as I can, in case anyone is still there who wants to talk to me, and I'll also be online tomorrow.

In the meantime, let me emphasize that I am OK with keeping things "as they are" for now (i.e. serialize fmpz/Int/BigInt/etc. separately -- my main goal was to make it explicitly what it is that we want. If that's the status quo and everyone can live with it for now, that's fine. On the long run, we may need something else, but that's a separate matter.

But we can discuss this more at some other point. Again, sorry for the short notice cancellation